MudListItem and MudMenuItem: Render as anchor with"Href" property (#1717)

belucha opened the pull request:
## Description

fixes #1717 

`MudElement` is used to render a MudListItem either using a `div` or an `a` element to render the item.
This way a list and or menu item with an Href can be opened in a new browser tab using the right click menu.
`MudElement` had to be modified the ClickPropagation property was depending on the HtmlTag value, this logic was moved to
the MudBaseButton.

## How Has This Been Tested?

- unit tests have been added
- visually

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [X] The PR is submitted to the correct branch (`dev`).
- [X] My code follows the code style of this project.
- [X] I've added relevant tests.

belucha committed:
fix #1717 MudListItem+MudMenuItem render as anchor, when "Href" prop is given

belucha commented:
The breaking change is in my opinion minor as it concerns `MudElement` with the `ClickPropagation` property not having internal logic.

belucha committed:
fix code quality: white space formatting

danielchalmers commented on src/MudBlazor/Components/Element/MudElement.cs:
```diff
@@ -38,7 +39,11 @@ public class MudElement : MudComponentBase
 
         [Parameter]
         [Category(CategoryTypes.Button.Behavior)]
-        public bool ClickPropagation { get; set; } = false;
+        public bool StopOnClickPropagation { get; set; }
```
I don't think this should be renamed when it's used in other controls as well, especially in the context of #6131

belucha replied:
I don't like this property either. The problem the name `ClickPropagation` is the opposite to `AddEventStopPropagationAttribute`.

In my opinion it would be best to completely remove the two Parameters and check, if it will work with attribute splatting  (UserAttributes). I will try that, since the implementation of `AddEventStopPropagationAttribute` only adds a special attribute.
```cs
public static void AddEventStopPropagationAttribute(this RenderTreeBuilder builder, int sequence, string eventName, bool value)
    {
        builder.AddAttribute(sequence, "__internal_stopPropagation_" + eventName, value);
    }
```
If that is true using it like this `<MudElement @onclick:stopPropagation="true"` might do all the work automatically.
I will try that. That would still be a breaking change, because the `ClickPropagation` parameter is removed. But I think this parameter was very rarely used and did only work if the HtmlTag was "button", which is totally not the expected behaviour.

```razor
@namespace MudBlazor
@inherits MudBaseButton

<MudElement @bind-Ref="@_elementReference"
            HtmlTag="@HtmlTag"
            Class="@Classname" 
            Style="@Style"
            @attributes="UserAttributes"
            @onclick="OnClickHandler"
            @onclick:stopPropagation="StopOnClickPropagation"
            type="@ButtonType.ToDescriptionString()"
            href="@Href" 
            target="@Target"
            rel="@GetRel()"
            disabled="@GetDisabledState()"
            >
    <span class="mud-button-label">
        @if (!string.IsNullOrWhiteSpace(StartIcon))
        {
            <span class="@StartIconClass">
                <MudIcon Icon="@StartIcon" Size="@(IconSize ?? Size)" Color="@IconColor" />
            </span>
        }
        @ChildContent
        @if (!string.IsNullOrWhiteSpace(EndIcon))
        {
            <span class="@EndIconClass">
                <MudIcon Icon="@EndIcon" Size="@(IconSize ?? Size)" Color="@IconColor" />
            </span>
        }
    </span>
</MudElement>
```

belucha replied:
Does not work the razor compiler does not accept this syntax on components:

```razor
...
            Style="@Style"
            @attributes="UserAttributes"
            @onclick="OnClickHandler"
            @onclick:stopPropagation="StopOnClickPropagation"
            type="@ButtonType.ToDescriptionString()"
...
```

```log
3>C:\github\intronik\MudBlazor\src\MudBlazor\Components\Button\MudFab.razor(9,23,9,37): error RZ10010: The component parameter 'onclick' is used two or more times for this component. Parameters must be unique (case-insensitive). The component parameter 'onclick' is generated by the '@onclick:stopPropagation' directive attribute.
3>C:\github\intronik\MudBlazor\src\MudBlazor\Components\Button\MudIconButton.razor(9,23,9,37): error RZ10010: The component parameter 'onclick' is used two or more times for this component. Parameters must be unique (case-insensitive). The component parameter 'onclick' is generated by the '@onclick:stopPropagation' directive attribute.
3>C:\github\intronik\MudBlazor\src\MudBlazor\Components\List\MudListItem.razor(8,23,8,42): error RZ10010: The component parameter 'onclick' is used two or more times for this component. Parameters must be unique (case-insensitive). The component parameter 'onclick' is generated by the '@onclick:stopPropagation' directive attribute.
3>C:\github\intronik\MudBlazor\src\MudBlazor\Components\List\MudListItem.razor(9,39,9,59): error RZ10010: The component parameter 'onclick' is used two or more times for this component. Parameters must be unique (case-insensitive). The component parameter 'onclick' is generated by the '@onclick:preventDefault' directive attribute.
```

belucha replied:
It is possible to omit the stop propagation parameter like this:

```razor
...
            @attributes="UserAttributes"
            @onclick="OnClickHandler"
            __internal_stopPropagation_onclick="@StopOnClickPropagation"
            type="@ButtonType.ToDescriptionString()"
            href="@Href" 
...
```

But it kind of relies on internal implementation of NET. So does the test currently as well.
Using this neither of the two parameters is required and all unit test and visual tests work.

henon replied:
We have parameters to prevent propagation and preventdefault on multiple components, so I wouldn't be concerned about having them on MudElement too. If we have them, they should be named consistently with what we usually use.

danielchalmers commented on src/MudBlazor/Components/Element/MudElement.cs:
```diff
@@ -38,7 +39,11 @@ public class MudElement : MudComponentBase
-        public bool ClickPropagation { get; set; } = false;
+        public bool StopOnClickPropagation { get; set; }
+
+        [Parameter]
+        [Category(CategoryTypes.Button.Behavior)]
+        public bool PreventOnClickDefault { get; set; }
```
Do you think we can be smarter about these types of properties instead of the user making the decision? I don't like that `ClickPropagation` exists either.

belucha replied:
See above, we should get rid of this.

danielchalmers commented on src/MudBlazor/Components/List/MudListItem.razor:
```diff
@@ -1,35 +1,49 @@
-
-<div tabindex="0" @attributes="UserAttributes" class="@Classname" @onclick="OnClickHandlerAsync" @onclick:stopPropagation="true" style="@Style">
-    @if (!string.IsNullOrWhiteSpace(Avatar)) {
+<MudElement HtmlTag="@HtmlTag"
```
nit: missing blank line between `@`s and `<`

belucha replied:
Sorry, I'm not getting it. What do you mean exactly?

danielchalmers replied:
Sorry, it is:

```razor
@typeparam T
<MudElement HtmlTag="@HtmlTag"
```

but it should be


```razor
@typeparam T

<MudElement HtmlTag="@HtmlTag"
```

belucha commented:
Another thing that I would like to discuss here.

What is the intended behaviour when both Href and OnClick handlers are present. I think the behaviour should be consistent across all components.

From my understanding, if Href is set the OnClick should not be called. But this decision imposes some troubles. It is for example not easily possible to add an MenuClose handler to inner MudListItem.
It  even gets more complex, if we add an Target attribute (e.g. _blank) and or a ForceLoad parameter. What is the expected behaviour for a disabled button/item with an href. Should the Href then also be disabled, by forcing a render as button tag.

For my expectation it would be best to always call the OnClick handler. That might close a menu and Browser client can do a navigation (even to a new tab) -- then the close menu is desired.

I think it is quite important to have a discussion on this **intended** behaviour, to get a consistent design!

My Pull-Request tried to solve the problem of #1717, but it seems there are more thoughts required!

ScarletKuro commented:
Sorry, I think my PR https://github.com/MudBlazor/MudBlazor/pull/8634 added conflicts to your PR, since I did method renaming, so was the tests affected too.

belucha committed:
Merge branch 'dev' into fix-1717-MudListItem-ShouldRenderAsAnchor

belucha committed:
MudElement: Keep ClickPropagation parameter and add PreventDefault Parameter

Previously the ClickPropagation parameter had MudElement internal logic. E.g. the parameter was only applied when the HtmlTag was an "button". This logic has been moved to MudBaseButton.

belucha commented:
I've merged with `dev` and kept the meaning of `MudElements` `ClickPropagation` Parameter.

danielchalmers commented on src/MudBlazor/Components/Element/MudElement.cs:
```diff
@@ -38,7 +39,11 @@ public class MudElement : MudComponentBase
+
+        [Parameter]
+        [Category(CategoryTypes.Button.Behavior)]
+        public bool PreventDefault { get; set; }
```
If this only applies to `onclick` the name should probably reflect that

belucha replied:
> If this only applies to onclick the name should probably reflect that

Yes it only applies to onclick, but other components use the same name too. In my original pull request I named it different `StopOnClickPropagation` someone argued, that I should keep the name.

We could name ist `OnClickPropagion` but the I liked my original proposal more, because it just uses the same names as dot net uses.

Honestly I'm very uncertain here.

danielchalmers commented on src/MudBlazor/Components/List/MudListItem.razor.cs:
```diff
@@ -267,7 +275,9 @@ internal void SetSelected(bool selected)
 
         private bool GetDisabled() => Disabled || (MudList?.GetDisabled() ?? false);
-
+        internal string HtmlTag => string.IsNullOrEmpty(Href) ? "div" : "a";
+        internal bool ApplyPreventDefault => GetDisabled();
+        internal bool ApplyClickPropagation => false;
```
Can these be not `internal`?

Should `ApplyClickPropagation` exist if it's always false?

danielchalmers commented on src/MudBlazor/Components/Menu/MudMenuItem.razor.cs:
```diff
@@ -25,12 +25,24 @@ public partial class MudMenuItem : MudComponentBase
+        [Category(CategoryTypes.Button.ClickAction)]
+        public string? Target { get; set; }
+
+        [Parameter]
+        [Category(CategoryTypes.Menu.ClickAction)]
+
+        public bool ForceLoad { get; set; }
```
I know it's not new but could you add a `<summary>` to `ForceLoad` as well so people understand what it's for?

belucha replied:
done

danielchalmers commented on src/MudBlazor/Base/MudBaseButton.cs:
```diff
@@ -137,6 +137,8 @@ private void SetDefaultValues()
 
         protected ElementReference _elementReference;
 
+        internal bool ApplyClickPropagation => !(HtmlTag == "button" && ClickPropagation == false);
```
Not `internal` if possible

danielchalmers replied:
`!(HtmlTag == "button" && ClickPropagation == false)` should be rewritten without the negation; something like `HtmlTag != "button" || ClickPropagation`

henon commented:
> What is the intended behaviour when both Href and OnClick handlers are present. I think the behaviour should be consistent 
> From my understanding, if Href is set the OnClick should not be called. But this decision imposes some troubles. It is for example not easily possible to add an MenuClose handler to inner MudListItem. It even gets more complex, if we add an Target attribute (e.g. _blank) and or a ForceLoad parameter. What is the expected behaviour for a disabled button/item with an href. Should the Href then also be disabled, by forcing a render as button tag.
> 
> For my expectation it would be best to always call the OnClick handler. That might close a menu and Browser client can do a navigation (even to a new tab) -- then the close menu is desired.
> 
> I think it is quite important to have a discussion on this **intended** behaviour, to get a consistent design!

I agree that calling the `OnClick` handler regardless of the `Href` value (except if `Disabled` is true) makes sense. If I attach a click handler I expect it to fire. If I want to just navigate I don't attach the handler, right?

danielchalmers commented:
> I agree that calling the `OnClick` handler regardless of the `Href` value (except if `Disabled` is true) makes sense. If I attach a click handler I expect it to fire. If I want to just navigate I don't attach the handler, right?

@henon Isn't it a problem that several other controls do it the other way?

henon commented:
@belucha Please resolve conflicts. What's the status? We should get this done for v7

belucha committed:
Merge remote-tracking branch 'MudBlazor/dev' into fix-1717-MudListItem-ShouldRenderAsAnchor

# Conflicts:
#	src/MudBlazor.UnitTests/Components/MenuTests.cs
#	src/MudBlazor/Base/MudBaseButton.cs
#	src/MudBlazor/Components/List/MudListItem.razor
#	src/MudBlazor/Components/List/MudListItem.razor.cs

belucha committed:
resove Formatting issues and adjust <summary> for Forceload

belucha committed:
remove `internal` modifier as requested by danielchalmers

belucha commented:
@henon I've just resolved all conflicts. 
@danielchalmers I also considered  your suggestions, except on regarding the rename of `MudElement.ClickPropagation`. I left that unchanged, since this was the original API and the the difference between OnClickPropagation and ClickPropagation is quite little.

henon commented on src/MudBlazor/Components/List/MudListItem.razor.cs:
```diff
@@ -322,6 +337,16 @@ private string GetIndeterminateIcon()
+        /// </summary>        
+        private string HtmlTag => string.IsNullOrEmpty(Href) || OnClickPreventDefault ? "div" : "a";
+
+        private bool ApplyPreventDefault => GetDisabled();
+
+        private bool ApplyClickPropagation => false;
```
Please change from private property to private method and rename `GetPreventDefault()` and `GetClickPropagation()` respectively.

henon commented on src/MudBlazor/Components/Menu/MudMenuItem.razor.cs:
```diff
@@ -89,23 +93,9 @@ protected async Task OnClickHandlerAsync(MouseEventArgs ev)
                 }
             }
 
-            if (Href != null)
-            {
-                if (string.IsNullOrWhiteSpace(Target))
-                {
-                    UriHelper.NavigateTo(Href, ForceLoad);
-                }
-                else
-                {
-                    await JsApiService.Open(Href, Target);
-                }
-            }
```
We don't do `UriHelper.NavigateTo(Href, ForceLoad);` or `JsApiService.Open(Href, Target);` any longer?

henon replied:
Ah, I guess the `MudListItem` does it.

henon reviewed (changes requested):
Just rename the private properties then it should be ready for merging.

belucha committed:
MudListItem: refactor two calc. properties to GetterFunctions()

belucha committed:
Merge remote-tracking branch 'MudBlazor/dev' into fix-1717-MudListItem-ShouldRenderAsAnchor

henon commented:
Thanks!

belucha commented:
@henon  Thanks to you and the others for the great work on this lib!

henon commented:
Added to v7.0.0 Migration Guide #8447

belucha commented:
@henon 
I'm so sorry, I found an issue in the code just now, that was not my intention and I somehow missed it.
in `MudBaseButton.cs` there is a property:

```cs
        public bool ApplyClickPropagation => HtmlTag != "button" || ClickPropagation;
```

this should be private and matching to your convention:

```cs
        private bool GetClickPropagation() => HtmlTag != "button" || ClickPropagation;
```

Do you wan't me to open a new pull request, or do you change that yourself?

Sorry for this mistake.

henon commented:
Don't worry, I am fixing it myself.

ScarletKuro commented:
After going through this PR again, I noticed that it was requested to add XML documentation for this property: https://github.com/MudBlazor/MudBlazor/pull/8649#discussion_r1567828070. 

However, it's missing:
https://github.com/MudBlazor/MudBlazor/blob/38782f5d83c481333526370ca274d8e45e389b32/src/MudBlazor/Components/Menu/MudMenuItem.razor.cs#L42-L46

Additionally, there is an extra new line and a missing new line after the property.

@belucha, can you please make a PR to fix it?

belucha commented:
Sure. I'm sorry if I missed that. I will do so in about two weeks.

ScarletKuro commented:
> Sure. I'm sorry if I missed that. I will do so in about two weeks.

Fixed already https://github.com/MudBlazor/MudBlazor/pull/9387
But if something is missing, feel free to PR.
